### PR TITLE
Include our own ClientHttpRequestFactory implementations

### LINKED ISSUE #17: Slow Reconnect with HttpComponents
If a client uses a `HttpComponentsClientHttpRequestFactory` re-connecting on processing failures (i.e.  `IOException`) is very slow. 

_fahrschein_ [tries to close](https://github.com/zalando-incubator/fahrschein/blob/master/src/main/java/org/zalando/fahrschein/NakadiReader.java#L190) the `ClientHttpResponse` in case of a failure. A `HttpComponentsClientHttpResponse` will always try to release the connection back to the connection manager [by consuming the full response body](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/client/HttpComponentsClientHttpResponse.java#L89) until an _EOF_. This can take an unpredictable amount of time, effectively pausing _fahrschein_.

See [here](https://github.com/lukasniemeier-zalando/fahrschein/blob/apache-blocks-on-close/src/test/java/org/zalando/fahrschein/NakadiReaderIT.java) for an example failing test.

### PR DESCRIPTION
**Pull request by jhorstmann**
To work around spring trying to drain the connection on close. This is a slightly less radical version of #160 

Fixes #17 and #83

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -151,11 +151,52 @@ Exception handling while streaming events follows some simple rules
...
 This library is currently tested and used in production with `SimpleClientHttpRequestFactory` and `HttpComponentsClientHttpRequestFactory`.
 
-Please note that `HttpComponentsClientHttpRequestFactory` and also `SimpleClientHttpRequestFactory` since spring 4.3.x try to consume the remaining stream on closing and so might block during reconnection.
+The `HttpComponentsClientHttpRequestFactory` and `SimpleClientHttpRequestFactory` of spring try to consume
+the remaining stream on closing and so might block during reconnection. To avoid this there are two alternative
+implementations included in the artifacts `fahrschein-http-simple` and `fahrschein-http-apache`.
+
+```java
+
+final RequestConfig config = RequestConfig.custom().setSocketTimeout(readTimeout)
```
Comment (lukasniemeier-zalando) [inline README.md]: Is this example showing the custom factories?

File: fahrschein-http-apache/src/main/java/org/zalando/fahrschein/http/apache/HttpComponentsClientHttpResponse.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	@Override
+	public InputStream getBody() throws IOException {
+		HttpEntity entity = this.httpResponse.getEntity();
+		return (entity != null ? entity.getContent() : new ByteArrayInputStream(new byte[0]));
+	}
+
+	@Override
+	public void close() {
+        // Release underlying connection back to the connection manager
+		if (this.httpResponse instanceof Closeable) {
```
Comment (lukasniemeier-zalando) [inline fahrschein-http-apache/src/main/java/org/zalando/fahrschein/http/apache/HttpComponentsClientHttpResponse.java]: formatting seems off here

File: README.md (hunk 1 above)
Comment (jhorstmann) [inline README.md]: Right, it's more showing the additional configuration possibilities of the apache impl. Also agree that there should be some guideline on when which impl should be used.

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Comment by fatroom** [comment]
@jhorstmann can you resolve conflicts in this PR?

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Comment by fatroom** [comment]
👍

**Pull request merged**